syncer/syncer.go

```diff
@@ -0,0 +1,402 @@
+	if oldState != newState {
+		s.logger.WithContext(ctx).With().Info("sync state change",
+			log.String("from", oldState.String()),
+			log.String("to", newState.String()),
+			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+			log.FieldNamed("latest", s.mesh.LatestLayer()),
+			log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+		events.ReportNodeStatusUpdate()
+		// notify subscribes
+		s.awaitSyncedMu.Lock()
+		defer s.awaitSyncedMu.Unlock()
+		if newState == synced {
+			close(s.awaitSyncedCh)
+		} else if oldState == synced {
+			s.awaitSyncedCh = make(chan struct{})
+		}
+	}
+}
+
+// startNewSyncProcess returns true if syncer is not currently synchronizing.
+func (s *Syncer) startNewSyncProcess() bool {
+	return atomic.CompareAndSwapUint32(&s.isBusy, 0, 1)
+}
+
+func (s *Syncer) endSyncProcess() {
+	atomic.StoreUint32(&s.isBusy, 0)
+}
+
+func (s *Syncer) setTargetSyncedLayer(ctx context.Context, layerID types.LayerID) {
+	newSyncLayer := uint64(layerID)
+	oldSyncLayer := atomic.SwapUint64((*uint64)(&s.targetSyncedLayer), newSyncLayer)
+	s.logger.WithContext(ctx).With().Info("target synced layer changed",
+		log.Uint64("from", oldSyncLayer),
+		log.Uint64("to", newSyncLayer),
+		log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+		log.FieldNamed("latest", s.mesh.LatestLayer()),
+		log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+}
+
+func (s *Syncer) getTargetSyncedLayer() types.LayerID {
+	return types.LayerID(atomic.LoadUint64((*uint64)(&s.targetSyncedLayer)))
+}
+
+func (s *Syncer) synchronize(ctx context.Context) bool {
+	logger := s.logger.WithContext(ctx)
+
+	if s.isClosed() {
+		logger.Warning("attempting to sync while shutting down")
+		return false
+	}
+
+	// at most one synchronize process can run at any time
+	if !s.startNewSyncProcess() {
+		logger.Info("sync is already running, giving up")
+		return false
+	}
+	// no need to worry about race condition for s.run. only one instance of synchronize can run at a time
+	s.run++
+	logger.Info("staring sync run #%v", s.run)
+
+	s.setStateBeforeSync(ctx)
+	// start a dedicated process for validation.
+	// do not use a unbuffered channel for vQueue. we don't want it to block if the receiver isn't ready. i.e.
+	// if validation for the last layer is still running
+	vQueue := make(chan *types.Layer, s.ticker.GetCurrentLayer())
+	vDone := make(chan struct{})
+	go s.startValidating(ctx, s.run, vQueue, vDone)
+	defer func() {
+		close(vQueue)
+		<-vDone
+		logger.With().Info("node is synced. validation done",
+			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+			log.FieldNamed("latest", s.mesh.LatestLayer()),
+			log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+		logger.Info("finished sync run #%v", s.run)
+		s.endSyncProcess()
+	}()
+
+	// using ProcessedLayer() instead of LatestLayer() so we can validate layers on a best-efforts basis.
+	// TODO: is it safe to skip layer 0
```

Comment:
our clock starts ticking from 1 so yes, you can skip layer 0

Reply:
thanks